Gnssrwnd1.0

KariA-Spire opened the pull request:
Description

Type of Change

New GNSS-R ocean wind speed observation operator and assimilation capability in GSI
This source code can directly assimilate GNSS-R L2 ocean wind speed retrievals from commercial Spire Global, Inc. and NASA/CYGNSS satellites

Resolves #628 

New feature (non-breaking change which adds functionality)
How Has This Been Tested?
Checklist

My code follows the style guidelines of this project
I have performed a self-review of my code
I have commented my code, particularly in hard-to-understand areas
Any dependent changes have been merged and published
File Changes

M [gsimain.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/gsimain.f90)

M [gsi_obOperTypeManager.F90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/gsi_obOperTypeManager.F90)

M[ intjo.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/intjo.f90)

M [_rhs.F90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/m_rhs.F90)

M [obsmod.F90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/obsmod.F90)

M [setuprhsall.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/setuprhsall.f90)

M [statsconv.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/statsconv.f90)

M [convinfo.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/convinfo.f90)

M [read_obs.F90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/read_obs.F90)

N [gsi_gnssrspdOper.F90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/gsi_gnssrspdOper.F90)

N [intgnssrspd.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/intgnssrspd.f90)

N [m_gnssrspdNode.F90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/m_gnssrspdNode.F90)

N [read_gnssrspd.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/read_gnssrspd.f90)

N[ setupgnssrspd.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/setupgnssrspd.f90)

N [stpgnssrspd.f90](https://github.com/NOAA-EMC/GSI/blob/bcc646efb6edf3a7786d3931959cc6b37a238e2b/src/gsi/stpgnssrspd.f90)

Link to the forked branch:https://github.com/KariA-Spire/GSI/tree/gnssrwnd1.0

Note: Upon request, we can facilitate source code for NCEP-flavor BUFR encoders, BUFR files, and JEDI/IODA converters. We can also guide users on observation error and bias correction treatment on an "as-time-permits" basis.

Spire Global, Inc. developed this assimilation capability under partial funding from the NASA CYGNSS ROSES-2020 Grant number 80NSSC21K1120.

KariA-Spire committed:
Add files via upload

Added the GNSS-R ocean wind speed assimilation to GSI. 
The following source code routines have been added or changed:
1.	gsi_gnssrspdOper.F90 
2.	intgnssrspd.f90
3.	m_gnssrspdNode.F90
4.	read_gnssrspd.f90
5.	setupgnssrspd.f90
6.	55 stpgnssrspd.f90
7.	23 gsimain.f90
8.	gsi_obOperTypeManager.F90
9.	intjo.f90
10.	rhs.F90
11.	obsmod.F90
12.	setuprhsall.f90
13.	statsconv.f90
14.	convinfo.f90
15.	read_obs.F90

KariA-Spire committed:
Add files via upload

KariA-Spire committed:
Add files via upload

Commit:
KA add GNSS-R ocean wind speed operator

Commit:
KA merged conflicts intjo, obsmod, setuprhsall, statsconv
Merge remote-tracking branch 'upstream/develop' into develop

Conflicts:
	src/gsi/intjo.f90
	src/gsi/obsmod.F90
	src/gsi/setuprhsall.f90
	src/gsi/statsconv.f90

Commit:
KA - add documentation to gnss-r spd routines

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd

KariA-Spire committed:
Add files via upload

Commit:
updated m_obsNodeTypeManager.F90

KariA-Spire committed:
Update obsmod.F90

Removed repeated block and merge conflicts

Commit:
Resolve merge conflicts

Commit:
Resolved merge conflicts

Commit:
obsmod.F90

Commit:
Addressed Russ's comments and updated documentation

Commit:
removed rhs_split_gps from m_rhs.F90

Commit:
Updates to statsconv.f90

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd1.0

emilyhcliu commented:
Great!!  This is exactly what we need.  
I will review it.

RussTreadon-NOAA commented:
Thank you @emilyhcliu

xincjin-NOAA commented on src/gsi/convinfo.f90:
```diff
@@ -242,9 +244,9 @@ subroutine convinfo_read
+       iotype='        '
        read(lunin,1030)cflg,iotype,crecord
-       if (cflg==' '.and.iotype=='       ') then
+       if (cflg=='  '.and.iotype=='        ') then
```
Any reasons to add a new space to compare with for cflg?

KariA-Spire replied:
As noted in the documentation section of convinfo_read, iotype was extended to 8 characters-long to allow for the character length of the new observation type **gnssrspd**.

xincjin-NOAA replied:
Was  the cflg extended to 2?

emilyhcliu replied:
Three questions about the observation type assignment
(1) Based on the code, both GNSS-R L2 and CYGNSS wind speed at 10 meters, correct?
(2) Are both (GNSS-R and CYGNSS)  assigned to GSI obstype as 298?
(3) Could you post the entries (parameters) for `gnssrspd` in convinfo here for our reference?

KariA-Spire replied:
@ xincjin-NOAA: thanks for your comment. The extra space in "cflg==" was a typo of merge conflict. convinfo.f90 has been updated. 

@emilyhcliu, here are the answers to your questions:

(1) Based on the code, both GNSS-R L2 and CYGNSS wind speed at 10 meters, correct? Yes
(2) Are both (GNSS-R and CYGNSS) assigned to GSI obstype as 298? Yes
(3) Could you post the entries (parameters) for gnssrspd in convinfo here for our reference?

![Screen Shot 2024-05-20 at 11 32 59 AM](https://github.com/NOAA-EMC/GSI/assets/104657787/82784130-6fd9-4d83-ba0d-cad9b07e65ee)

xincjin-NOAA commented on src/gsi/m_obsNodeTypeManager.F90:
```diff
@@ -125,7 +128,7 @@ module m_obsNodeTypeManager
   public:: iobsNode_light
   public:: iobsNode_dbz
-  public:: iobsNode_fed
+  public:: iobsNode_gnssrspd
```
I saw all the iobsNode_fed were removed from the code base. is the iobsNode_gnssrspd supposed to replace the iobsNode_fed?

KariA-Spire replied:
This was a merge conflict. All "fed" instances were added to m_obsNodeTypeManager.F90.

xincjin-NOAA replied:
Thanks

KariA-Spire replied:
Thank you @xincjin-NOAA!

emilyhcliu commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,514 @@
+     maxobs = 0
+     call closbf(lunin) 
+     open(lunin,file=trim(infile),form='unformatted')
+     call openbf(lunin,'IN',lunin)
```
@KariA-Spire The Input data is BUFR format.  Do you bufrize the data yourself or it is done by NESDIS?

KariA-Spire replied:
We do bufrize the data.

emilyhcliu commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,957 @@
+!    if ( nty == 260 .or. nty == 261) z_height = .true.
+!    nty == 213 is temporarily assigned to SFMR retrieved wind speed from recon
+!    and is subjet to change in the future
+     if ( nty == 260 .or. nty == 261 .or. nty == 213) z_height = .true.
```
Do we need this here for wind speed from GNSS-R/CYGNSS?

KariA-Spire replied:
This is legacy code. Lines 328-331 were removed from setupgnssrspd.f90

emilyhcliu commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,514 @@
+              cdata_all( 1,iout)=woe                    ! wind error
+              cdata_all( 2,iout)=dlon                   ! grid relative longitude             
+              cdata_all( 3,iout)=dlat                   ! grid relative latitude                  
+              cdata_all( 4,iout)=dlnpsob                ! ln(surface pressure in cb)
```
`dlnpsob` is declared but not defined (no given value)

KariA-Spire replied:
@emilyhcliu: this line was left in case some type of pressure information gets included in the BUFR files in the future. I can remove it if you think it's best.

RussTreadon-NOAA replied:
If the line `cdata_all( 4,iout)=dlnpsob` is retained, we must provide a default value for `dlnpsob`.  Otherwise `cdata_all(4,iout)` will be loaded with an uninitialized value.  This is unsafe coding.

KariA-Spire replied:
Done!

emilyhcliu commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,957 @@
+        presw    = ten*exp(pobl)
+
+!    Process observations with reported pressure
+     else
```
This setup routine processes the wind speed from GNSS, which is 10-meter wind speed. Therefore, z_height is always true. Do we need the section in which the GNSS wind speeds come with the reported pressure level (as metadata)? Could GNSS wind speed data come with pressure in the future?

KariA-Spire replied:
@emilyhcliu same comment as obove. Thanks.

RussTreadon-NOAA replied:
@emilyhcliu , may this conversation be marked as resolved?  If "yes", please do so.  If "no", what additional information do you need?

KariA-Spire committed:
Update convinfo.f90: removed extra space in cflg

xincjin-NOAA commented:
In general I am fine with this PR. One question in my mind is that how can I know if the logic in the codes are correct. With so many lines of codes, it is vary hard to judge it by just reading them. Do we have some kind of test cases available some where?

KariA-Spire committed:
Update setupgnssrspd.f90

Removed legacy code.

KariA-Spire committed:
Update m_obsNodeTypeManager.F90

Reconciles merge conflicts wrt fed.

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd1.0

emilyhcliu commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,514 @@
+              cdata_all( 2,iout)=dlon                   ! grid relative longitude             
+              cdata_all( 3,iout)=dlat                   ! grid relative latitude                  
+              cdata_all( 4,iout)=dlnpsob                ! ln(surface pressure in cb)
+              cdata_all( 5,iout)=spdob*sqrt(two)*half   ! u obs
+              cdata_all( 6,iout)=spdob*sqrt(two)*half   ! v obs
```
@KariA-Spire 
The observation is `wind speed` (spdob).
The u-wind (uob) and v-wind (vob) components are derived by assuming uob = vob.
And, the uob and vob will always be positive.  
I am curious: what is the scientific basis for this assumption?

KariA-Spire replied:
As a first approximation for assimilating GNSS-R wind speed observations, some members of the CYGNSS science team suggested using existing GSI code for reading ocean surface wind speed from a Step Frequency Microwave Radiometer (SFMR).  There are multiple well-known meteorological scenarios where the zonal and meridional wind components are equal in magnitude at the ocean surface. For example, in the inter-tropical convergence zone trade winds from the subtropics converge with winds originating from the opposite direction, the resulting wind field can have a 45-degree angle with respect to the u and v components, making them equal in magnitude. Similarly, during tropical cycles, storm displacement can lead to diagonal wind fields relative to the zonal and meridional wind components. Analogous to SFMR wind speed observations, GNSS-R observations can benefit tropical weather observation and prediction. Moreover, this calculation was chosen while the GNSS-R community continues to develop and test a wind direction retrieval algorithm or a machine-learned estimation to allow for a more precise remotely sensed measurement of the zonal and meridional wind components.

emilyhcliu replied:
@KariA-Spire Thanks for your explanation and the slides.  
I understand that the assimilation of GNSS-R wind speed is following the footstep of FSMR retrieved wind speed, which also assumes uob and vob.  I believe that must be QC associated with this assumption for FSMR in GSI.

Thanks for your clarification.  It is good to hear that the wind direction retrieval is in progress.  
I am good with your changes.

KariA-Spire replied:
@emilyhcliu, thank you very much for reviewing this PR and for your insightful comments.

RussTreadon-NOAA replied:
@emilyhcliu , if this conversation is resolved, please mark it as so.  We can not process GSI PRs with unresolved coversations.

KariA-Spire committed:
Update read_gnssrspd.f90

KariA-Spire commented:
@xincjin-NOAA thank you for your comment: "In general I am fine with this PR. One question in my mind is that how can I know if the logic in the codes are correct. With so many lines of codes, it is vary hard to judge it by just reading them. Do we have some kind of test cases available some where?"

I’m including a link to a presentation that includes some results from a case study with CYGNSS GNSS-R ocean wind speed observations in the tropics:

https://www.ssec.wisc.edu/meetings/iwwg/2023-meeting/presentations/assessing-the-benefit-of-assimilating-cygnss-ocean-reflectometry-delay-doppler-maps-in-a-4denvar-system-for-improving-global-operational-analysis-and-forecasting/

We are in the process of preparing a final report for NASA and a peer-reviewed publication that will be finalized in the next few months. We'd be happy to share those after the appropriate reviews. Regarding caste studies using Spire proprietary observations, Spire’s legal team has yet to vet the results before we can publicly release them.

emilyhcliu reviewed (approved):
My questions have been addressed.  Approved!!

xincjin-NOAA commented:
@KariA-Spire Thank you for your response which is very helpful.

xincjin-NOAA reviewed (approved):
The changes looks good to me.

RussTreadon-NOAA commented:
**WCOSS2 test**
Install `KariA-Spire:gnssrwnd1.0` at 9ba3f84f and `develop` at 59d7578b on Cactus.  Run ctests (regression) with following results
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr747/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............***Failed  486.77 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  489.33 sec
3/7 Test #7: global_enkf ......................   Passed  858.16 sec
4/7 Test #2: rtma .............................   Passed  968.89 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1156.42 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1213.35 sec
7/7 Test #1: global_4denvar ...................   Passed  1747.67 sec

86% tests passed, 1 tests failed out of 7

Total Test time (real) = 1747.68 sec

The following tests FAILED:
          4 - netcdf_fv3_regional (Failed)
```

`netcdf_fv3_regional` failed the memory threshold test.
```
The memory for netcdf_fv3_regional_loproc_updat is 219428 KBs.  
This has exceeded maximum allowable memory of 179986 KBs, resulting in Failure memthresh of the regression test.
```
The memory threshold test is not a reliable measure of the memory used by `gsi.x`.  This is not a fatal fail.

Note that none of the ctests exercise the functionality added by this PR.  As such, the _Passed_ result is expected.

RussTreadon-NOAA commented on src/gsi/obsmod.F90:
```diff
@@ -767,6 +768,7 @@ subroutine init_obsmod_dflts
     if_vterminal=.false.
     l2rwthin    =.false.  
     if_vrobs_raw=.false.
+    if_model_dbz=.true.
```
Why is this line added here?   `if_model_dbz` is already set to `.false`. on line 773.   Please remove line 771.

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/gsimain.f90:
```diff
@@ -1,6 +1,7 @@
 !-------------------------------------------------------------------------
 !  NASA/GSFC, Global Modeling and Assimilation Office, Code 610.3, GMAO  !
 !-------------------------------------------------------------------------
+
```
Please remove this added blank line.

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/gsi_obOperTypeManager.F90:
```diff
@@ -280,20 +286,18 @@ function dtype2index_(dtype) result(index_)
     case("ompstc8"); index_= iobOper_oz
     case("ompsnp" ); index_= iobOper_oz
     case("ompsnm" ); index_= iobOper_oz
-    case("omieff"   ); index_= iobOper_oz
```
Why are there changes in the ozone section of this file?

KariA-Spire replied:
Updated this section to match the Develop branch.

RussTreadon-NOAA replied:
Differences remain between `develop` and `KariA-Spire:gnssrwnd1.0` with regards to ozone. 

`develop:gsi_obOperTypeManager.F90` [contains](https://github.com/NOAA-EMC/GSI/blob/59d7578b31454140cb38bf65b27e3cffb02c7e3e/src/gsi/gsi_obOperTypeManager.F90#L276-#L296) 
```
  case("oz"     ,"[ozoper]"     ); index_= iobOper_oz
    case("sbuv2"  ); index_= iobOper_oz
    case("omi"    ); index_= iobOper_oz
    case("gome"   ); index_= iobOper_oz
    case("ompstc8"); index_= iobOper_oz
    case("ompsnp" ); index_= iobOper_oz
    case("ompsnm" ); index_= iobOper_oz
    case("omieff"   ); index_= iobOper_oz
    case("tomseff"  ); index_= iobOper_oz
    case("ompsnmeff"); index_= iobOper_oz

  case("o3l"    ,"[o3loper]"    ); index_= iobOper_o3l
    case("o3lev"    ); index_= iobOper_o3l
    case("mls20"    ); index_= iobOper_o3l
    case("mls22"    ); index_= iobOper_o3l
    case("mls30"    ); index_= iobOper_o3l
    case("mls55"    ); index_= iobOper_o3l
    case("ompslp"   ); index_= iobOper_o3l
    case("ompslpuv" ); index_= iobOper_o3l
    case("ompslpvis"); index_= iobOper_o3l
    case("ompslpnc" ); index_= iobOper_o3l
```
whereas `KariA-Spire:gnssrwnd1.0:gsi_obOperTypeManager` [contains](https://github.com/KariA-Spire/GSI/blob/2aba9bc75fc4ee120e122a7ac0863f4851ee26b0/src/gsi/gsi_obOperTypeManager.F90#L282-#L304)
```
  case("o3l"    ,"[o3loper]"    ); index_= iobOper_o3l
    case("o3lev"    ); index_= iobOper_o3l
    case("mls20"    ); index_= iobOper_o3l
    case("mls22"    ); index_= iobOper_o3l
    case("mls30"    ); index_= iobOper_o3l
    case("mls55"    ); index_= iobOper_o3l
    case("ompslp"   ); index_= iobOper_o3l
    case("ompslpuv" ); index_= iobOper_o3l
    case("ompslpvis"); index_= iobOper_o3l
    case("ompslpnc" ); index_= iobOper_o3l

  case("o3l"    ,"[o3loper]"    ); index_= iobOper_o3l
    case("o3lev"    ); index_= iobOper_o3l
    case("mls20"    ); index_= iobOper_o3l
    case("mls22"    ); index_= iobOper_o3l
    case("mls30"    ); index_= iobOper_o3l
    case("mls55"    ); index_= iobOper_o3l

    case("omieff"   ); index_= iobOper_o3l
    case("tomseff"  ); index_= iobOper_o3l
    case("ompslpuv" ); index_= iobOper_o3l
    case("ompslpvis"); index_= iobOper_o3l
    case("ompslp"   ); index_= iobOper_o3l
```
`KariA-Spire:gnssrwnd1.0:gsi_obOperTypeManager` contains two sections for `case("o3l"`

RussTreadon-NOAA replied:
Ozone differences still remain with respect to `develop`

KariA-Spire replied:
Corrected "oz" and "o3loper" blocks to match develop branch

RussTreadon-NOAA commented on src/gsi/gsi_obOperTypeManager.F90:
```diff
@@ -280,20 +286,18 @@ function dtype2index_(dtype) result(index_)
     case("mls30"    ); index_= iobOper_o3l
     case("mls55"    ); index_= iobOper_o3l
-    case("ompslp"   ); index_= iobOper_o3l
+    case("omieff"   ); index_= iobOper_o3l
```
Why are there changes in the ozone section of this file?

KariA-Spire replied:
Updated this section to match the Develop branch.

RussTreadon-NOAA replied:
See comment above

KariA-Spire replied:
Corrected "oz" and "o3loper" blocks to match develop branch

RussTreadon-NOAA commented on src/gsi/gsi_obOperTypeManager.F90:
```diff
@@ -363,7 +367,6 @@ function dtype2index_(dtype) result(index_)
   case("aero"   ,"[aerooper]"   ); index_= iobOper_aero
     case("aod"      ); index_= iobOper_aero
     case("modis_aod"); index_= iobOper_aero
-    case("viirs_aod"); index_= iobOper_aero
```
Why is this line removed?

KariA-Spire replied:
Updated this section to match the Develop branch.

RussTreadon-NOAA commented on src/gsi/intjo.f90:
```diff
@@ -255,6 +258,7 @@ subroutine intjo_(rval,qpred,sval,sbias)
 class(obOper),pointer:: it_obOper
 
 !******************************************************************************
+  call setrad(sval(1))
```
Why is this call added?

KariA-Spire replied:
Removed.

RussTreadon-NOAA replied:
Remove `use intradmod, only: setrad` since it is no longer needed.

KariA-Spire replied:
Done!

RussTreadon-NOAA replied:
Thank you.

RussTreadon-NOAA commented on src/gsi/m_obsNodeTypeManager.F90:
```diff
@@ -122,10 +126,12 @@ module m_obsNodeTypeManager
   public:: iobsNode_swcp
   public:: iobsNode_lwcp
-
+ 
```
Please remove extra whitespace

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/m_obsNodeTypeManager.F90:
```diff
@@ -250,6 +259,7 @@ module m_obsNodeTypeManager
     enumerator:: iobsNode_dbz
     enumerator:: iobsNode_fed
 
+
```
Remove extra blank line.

KariA-Spire replied:
Done!

RussTreadon-NOAA replied:
The extra blank [line](https://github.com/KariA-Spire/GSI/blob/2aba9bc75fc4ee120e122a7ac0863f4851ee26b0/src/gsi/m_obsNodeTypeManager.F90#L260) remains.

RussTreadon-NOAA replied:
The extra blank line remains.

KariA-Spire replied:
Removed!

RussTreadon-NOAA commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,515 @@
+     ncsave  =  0
+     do nc = 1, nconvtype
+               if (trim(ioctype(nc)) == trim(obstype)) then 
+                  if (trim(ioctype(nc)) == 'gnssrspd' .and. ictype(nc) == 298 ) then
```
We need to be careful here.  [RTMA](https://www.emc.ncep.noaa.gov/emc/pages/infrastructure/obs-data/prepbufr.doc/table_19.php) uses 298 for a different data type.  We should coordinate with the ObsProc team to select a unique identifier for this data type.

KariA-Spire replied:
@RussTreadon-NOAA, thanks for the suggestion. 
@ilianagenkova, could you please suggest a number for a new observation type (GNSS-R wind speed) that is not used for any other model?

ilianagenkova replied:
@RussTreadon-NOAA and @KariA-Spire , assigning an obs type  (a.k.a **kx** in GSI) needs to be done carefully. If the GNSS-R wind speed will never be added to the prepbufr files ( I recommend it!) then any number >= 600 and <999 would be a good one.  

Why that range? - values up to 299 are already used, values 300-599 appear in GSI's read_obs.F90 so we better not mess with these.

I suggest GNSS-R wind speed is kx=600. If you agree, tag me again, so I record it somewhere in the ObsProc documents. 

The [RTMA table](https://www.emc.ncep.noaa.gov/emc/pages/infrastructure/obs-data/prepbufr.doc/table_19.php) Russ provided covers all kx values that come from obsproc, and more specifically via the prepbufr generation. 

However, if a data set is not going in the prepbufr files, the kx could be assigned in the GSI code and used in the global_convinfo.txt. For example, we assigned kx=999 to the short-lived "dw" Aeolus data, as the table below show:
![image](https://github.com/NOAA-EMC/GSI/assets/59100254/c2a3548a-8dcc-48fe-ba6c-b3e85a55224e)
and we ran GSI experiments without coordiation with ObsProc. (we later assigned kx=48 value to Aeolus, which is again - outside of the 100-299 range "occupied" by prepbufr data types.

@RussTreadon-NOAA , if you know about the origin of the kx=300-599 values in read_obs.F90, please let me know.

KariA-Spire replied:
Thank you @ilianagenkova for the detailed explanation. I agree with assigning kx=600 to GNSS-R wind speed.

RussTreadon-NOAA replied:
What are the bufr satellite identifier(s) for the satellite(s) from which the gnss-r data is derived?

emilyhcliu replied:
> We need to be careful here. [RTMA](https://www.emc.ncep.noaa.gov/emc/pages/infrastructure/obs-data/prepbufr.doc/table_19.php) uses 298 for a different data type. We should coordinate with the ObsProc team to select a unique identifier for this data type.

Did not spot this!!  kx=298 is taken.

emilyhcliu replied:
> What are the bufr satellite identifier(s) for the satellite(s) from which the gnss-r data is derived?

@KariA-Spire, I could not find the satellite identifier for GNSS-R and CYGNSS. The BUFR data usually includes this information in a mnemonic called `SATID`.

KariA-Spire replied:
@emilyhcliu  and @RussTreadon-NOAA we did not add any satellite identifiers when we created our own bufr files.

RussTreadon-NOAA commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,515 @@
+              spdob = gnssrw(1,1) ! surface wind speed 
+
+              ! Don't permit observations with ws <= 1 m/s
+              if (spdob <= 1.0) cycle loop_readsb2
```
Append `_r_kind` to `1.0` or use constant `one`

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,515 @@
+
+              ! Get observation error from bufr file
+              call ufbint(lunin,gnssrw,1,1,nlv,oestr)
+              obserr = max(gnssrw(1,1),1.5) ! surface wind speed error
```
append `1.5` with `_r_kind`

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/read_obs.F90:
```diff
@@ -487,6 +490,15 @@ subroutine read_obs_check (lexist,filename,jsatid,dtype,minuse,nread)
+               exit gnssrwndloop
+           endif
+         end do gnssrwndloop
+
```
Seems we should have the line `nread = nread + 1` here, right?

KariA-Spire replied:
We didn't seem to need this line, but I added it.

RussTreadon-NOAA commented on src/gsi/read_obs.F90:
```diff
@@ -1449,8 +1461,14 @@ subroutine read_obs(ndata,mype)
                 endif
+
+             else if (obstype == 'gnssrspd' .and. index(infile,'gnssrwndbufr') /=0 ) then
+                write(6,*) " Before read_gnssrspd: ndat=",ndat," nread=",nread," npuse=",npuse," nouse=",nouse
```
This write statement seems to be leftover from initial code development or debugging.  Please remove this line.

KariA-Spire replied:
Removed.

RussTreadon-NOAA commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,951 @@
+  integer(i_kind)                                  ,intent(in   ) :: is ! ndat index
+
+! Declare local variables
+  real(r_kind),parameter:: ten=10.0_r_kind
```
`constants.f90` defines `ten`.  Remove this line and add `ten` to the `use constants` line above.

KariA-Spire replied:
Thanks for the suggestion!

RussTreadon-NOAA commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,951 @@
+!    Process cygnss observations at mslp
+     if ( nty == 298) then
+          z_height = .true.
+          data(ihgt,i) = 10
```
Replace `10` with the constant `ten`

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,515 @@
+     ntmatch =  0
+     ncsave  =  0
+     do nc = 1, nconvtype
+               if (trim(ioctype(nc)) == trim(obstype)) then 
```
if blocks in this section of code need to be aligned properly

KariA-Spire replied:
Done!

RussTreadon-NOAA commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,951 @@
+
+!    Process cygnss observations at mslp
+     if ( nty == 298) then
+          z_height = .true.
```
`z_height` is not defined if `nty /= 298`.  It is dangerous to assume that unfinalized logical variables default to `.false.`   Provide a default value for `z_height`

KariA-Spire replied:
I initialized z_height to false. Good catch!

RussTreadon-NOAA commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,951 @@
+        presw = ten*exp(dpres)
+        dpres = dpres-log(psges)
+        drpx=zero
+        if(nty >= 280 .and. nty < 290)then
```
I thought from `read_gnssrspd.f90` that the obs type was 298.  Can it also fall in the range [280, 290)?

KariA-Spire replied:
I left this part of the code in case the finalized BUFR files end up containing GNSS-R wind observations with a reported pressure value.

RussTreadon-NOAA replied:
If the final BUFR files contain GNSS-R wind observations with  reported pressure value, wouldn't the type still be 298?

`setupgnssrspd.f90` has logic for `nty == 298` and `nty >= 280 .and. nty < 290`.  This doesn't completely align with `nty = 298` in `read_gnssrspd.f90`.   This is confusing.

KariA-Spire replied:
The numbers will have to be revisited or I could just delete this part for now.

RussTreadon-NOAA replied:
Shouldn't we finalize the `nty` values before we write code?  Doing so will likely also impact `read_gnssrspd.f90`.  It's odd to write code with specific type numbers when we don't know if these are the type numbers we'll actually use.

KariA-Spire replied:
I will delete this section. I don't know if NOAA will eventually have this information in their bufr files.

KariA-Spire replied:
@ilianagenkova FYI, in case obsproc ends up processing GNSS-R bufr files with reported surface pressure information, nty types will have to be assigned and setupgnssrspd.f90 and read_gnssrspd.f90 will need to be updated.

ilianagenkova replied:
@KariA-Spire, a few questions: 
1. Is "nty" serving as "kx" for GNSS-R?
2. How is the presence of reported surface pressure relevant to / impacting "nty"?
3. What is the nty value at the moment (in any of the  *gnssrspd*.f90)
4. A word of caution to everyone (@RussTreadon-NOAA @ADCollard @emilyhcliu)- while every conventional data set (data types listed in global_convinfo.txt) has a "kx" in the global_convinfo.txt and that "kx" is used across the GSI code, not all of these "kx" are know/set in the ObsProc code. As long as a "kx" that is introduced in GSI isn't  overlapping with the "kx" assigned within OBsProc (upastream of GSI) there should be no problems.

KariA-Spire replied:
@ilianagenkova 
Answers to your questions:
1. Yes
2. If there is a range of observations with specific reports of surface pressure, they would have to be assigned to nty or kx range
3. I changed it to 600 and also in read_gnssrspd.f90 per your ealier recommendation.: https://github.com/NOAA-EMC/GSI/pull/747#discussion_r1626645923

RussTreadon-NOAA commented on src/gsi/statsconv.f90:
```diff
@@ -676,18 +678,46 @@ subroutine statsconv(mype,&
+  end if
+! Summary report for conventional td2m
+  if(mype==mype_td2m) then
         if(first)then
```
The code indentation in this block is not correct.  Please properly align code blocks for improved readability.

KariA-Spire replied:
Revised!

KariA-Spire replied:
Revised!

RussTreadon-NOAA commented on src/gsi/statsconv.f90:
```diff
@@ -676,18 +678,46 @@ subroutine statsconv(mype,&
         close(iout_wspd10m)
      end if
   end if
+! Summary report for conventional gnssrspd
```
The order in which things are done in this block differs from the processing of other obs.  Is there a reason for this?

KariA-Spire replied:
This section follows the order of 2dtm in the same file. I did notice that statsconv.f90 in the develop branch has been updated. I will match the develop branch

RussTreadon-NOAA reviewed (commented):
Look at code from coding standards perspective.  Please see questions & requests for changes.

KariA-Spire committed:
Update obsmod.F90

Removed line 771: if_model_dbz=.true.

KariA-Spire committed:
Update gsimain.f90

Removed blank line number 4.

KariA-Spire committed:
Update gsi_obOperTypeManager.F90

Changed the ozone section to match the Develop branch.

KariA-Spire committed:
Update gsi_obOperTypeManager.F90

Updated the Ozone section to match the Develop branch.

KariA-Spire committed:
Update intjo.f90

Removed extra call:   call setrad(sval(1))

KariA-Spire committed:
Update m_obsNodeTypeManager.F90

Removed extra blank space.

RussTreadon-NOAA commented on src/gsi/intjo.f90:
```diff
@@ -240,6 +241,8 @@ subroutine intjo_(rval,qpred,sval,sbias)
 use m_obsdiags, only: obOper_destroy
 use gsi_obOper, only: obOper
 
+use intradmod, only: setrad
```
Since `setrad(sval(1))` has been removed the `use intradmod` line is no longer needed.  Please remove this line and the blank line following it.

RussTreadon-NOAA commented:
Thank you @KariA-Spire for addressing some of my comments.  I resolved _conversations_ which are resolved.  I'll review and resolve other _conversations_ as they are addressed.

KariA-Spire committed:
Update gsi_obOperTypeManager.F90

KariA-Spire committed:
Update gsi_obOperTypeManager.F90

Corrected "oz" and "o3loper" blocks to match develop branch

KariA-Spire committed:
Update m_obsNodeTypeManager.F90

KariA-Spire committed:
Update read_gnssrspd.f90

L415: used constant one
L420: append 1.5 with _r_kind

KariA-Spire committed:
Update intjo.f90

Removed use intradmod, only: setrad

RussTreadon-NOAA commented:
Please update `KariA-Spire:gnssrwnd1.0` to the current head of NOAA-EMC/GSI `develop`.  `gnssrwnd1.0` is 4 commits behind the current head of `develop`.

RussTreadon-NOAA commented on src/gsi/setupgnssrspd.f90:
```diff
@@ -0,0 +1,951 @@
+  ! apply only to the regional forecast models (e.g., HWRF); Henry
+  ! R. Winterbottom (henry.winterbottom@noaa.gov).
+  
+  use obsmod, only: uv_doe_a_213,uv_doe_b_213
```
Please remove this line.  `uv_doe_a_213` and `uv_doe_b_213` are not used in this file.

KariA-Spire replied:
Done!

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd1.0

KariA-Spire committed:
Update read_obs.F90

KariA-Spire committed:
Update setupgnssrspd.f90

KariA-Spire committed:
Update read_gnssrspd.f90

Changed  to ictype(nc) == 600

KariA-Spire committed:
Update setupgnssrspd.f90

Changed nty == 600

KariA-Spire committed:
Update statsconv.f90

KariA-Spire committed:
Update read_gnssrspd.f90

KariA-Spire committed:
Update statsconv.f90

KariA-Spire committed:
Update setupgnssrspd.f90

KariA-Spire committed:
Update setupgnssrspd.f90

KariA-Spire committed:
Update setupgnssrspd.f90

Removed this block. 
!    Process observations with reported pressure
     else

        presw = ten*exp(dpres)
        dpres = dpres-log(psges)
        drpx=zero
        if(nty >= 280 .and. nty < 290)then
           dpresave=dpres
           dpres=-goverrd*data(ihgt,i)/tges(1)
           if(nty < 283)drpx=abs(dpres-dpresave)*factw*thirty
        end if
Might need to be added in the future if gnss-r l2 wind speed retrievals include observations with reported pressures.

KariA-Spire committed:
Update read_gnssrspd.f90

RussTreadon-NOAA commented:
Thank you @KariA-Spire for recent updates.  Please update `KariA-Spire:gnssrwnd1.0` with the current head of `NOAA-EMC:develop`.  The last commit to `develop` enabled the Orion Rocky 9 build.  We want to test your branch on Orion.

RussTreadon-NOAA commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,519 @@
+     real(r_kind) :: dx,dy,dx1,dy1,w00,w10,w01,w11
+     real(r_kind) :: wdir,wspd
+     real(r_kind) :: tob,uob,vob,qob,spdob,rrob
+     real(r_kind) :: rhob,tdob,dlnpsob,log
```
We don't need to declare `log`.  This is a fortran intrinsic function.

KariA-Spire replied:
Removed!

RussTreadon-NOAA commented on src/gsi/read_gnssrspd.f90:
```diff
@@ -0,0 +1,519 @@
+     real(r_kind) :: pob_mb,pob_cb,pob_pa,gob
+     real(r_kind) :: psob_mb,psob_cb,psob_pa
+     real(r_kind) :: qmaxerr 
+     real(r_kind) :: dlnpsob,dlnpob,ppb
```
`dlnpsob` was declared on line 157.  Either remove the `dlnpsob` declaration on this line or line 157.  The code does not compile with two declarations of the same variable.  

Notice that `ppb` is not used in this routine.  Please check the routines you added or modified and remove unused variables.

KariA-Spire replied:
Removed!

RussTreadon-NOAA commented:
**WCOSS2 (Dogwood) ctest**
Build `KariA-Spire:gnssrwnd1.0` at 8ad5c74b.  Remove double declaration of `dlnpsob` in `src/gsi/read_gnssrspd.f90`.  Run ctests with following results
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr747/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  731.17 sec
2/6 Test #6: global_enkf ......................   Passed  918.48 sec
3/6 Test #2: rtma .............................   Passed  1032.55 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  1215.66 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  1274.36 sec
6/6 Test #1: global_4denvar ...................   Passed  1743.76 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1744.04 sec
```

KariA-Spire committed:
Update read_gnssrspd.f90

KariA-Spire committed:
Merge branch 'NOAA-EMC:develop' into gnssrwnd1.0

KariA-Spire commented:
@RussTreadon-NOAA, I want to give you a big THANK YOU for the thorough review. I believe I have addressed all of your latest comments, but please let me know if you have any additional concerns and the next steps.

Also, thank you very much @emilyhcliu, @xincjin-NOAA, and @ilianagenkova for your reviews and comments. 
 
Best regards!
Kari

xincjin-NOAA commented:
@KariA-Spire Could you send me an example BUFR file of the data so that I can make a test run for all these codes? In addition,  I am wondering if you could send me the source code for NCEP-flavor BUFR encoders?  I am suppose to assimilate the GNSS_R wind speed in GSI, the source code will be very helpful.  Thanks!

Xin

KariA-Spire commented:
@xincjin-NOAA, could you please send me an email with your request so I can forward it to the appropriate individuals at Spire.  Karina . Apodaca @ spire . com

xincjin-NOAA commented:
> @xincjin-NOAA, could you please send me an email with your request so I can forward it to the appropriate individuals at Spire. Karina . Apodaca @ spire . com

Thank you so much! I have sent an email request to the email above.

Xin

RussTreadon-NOAA commented:
Thank you @xincjin-NOAA for your comment and request.  I'm glad to hear that you will test the code with data.  None of our existing GSI regression tests exercise the new code.  

If it does not take too long to obtain data and run run GNSS_R wind data through the code, we will keep this PR open until tests confirm that the code works on WCOSS2 and/or NOAA RDHPCS machines as expected.

RussTreadon-NOAA commented:
**Hercules ctest**
Install `KariA-Spire:gnssrwnd1.0` at 16a0149a on Hercules.  Run ctests with following results
```
Test project /work/noaa/da/rtreadon/git/gsi/pr747/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  545.52 sec
2/6 Test #6: global_enkf ......................   Passed  906.54 sec
3/6 Test #2: rtma .............................   Passed  1264.87 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  1392.48 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  1457.69 sec
6/6 Test #1: global_4denvar ...................   Passed  1680.99 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1681.00 sec
```

ilianagenkova commented:
@KariA-Spire , where from did you read the GNSS-R data you assimilated?  In NCEP at the moment we have *gpsro* and *gpsipw*, but as far as I know they don't contain wind info.

RussTreadon-NOAA commented:
GSI `develop` will become the GSI package included in GFS v17.  NCO compiles codes in debug mode with a flag for unused variables.  Unused variables must be removed from implementation code.  The following source code files in `KariA-Spire:gnssrwnd1.0` contain unused variables
```
src/gsi/gsi_gnssrspdOper.F90(63): remark #7712: This variable has not been used.   [INIT_PASS]
src/gsi/gsi_gnssrspdOper.F90(63): remark #7712: This variable has not been used.   [LAST_PASS]
src/gsi/read_gnssrspd.f90(1): remark #7712: This variable has not been used.   [GSTIME]
src/gsi/read_gnssrspd.f90(41): remark #7712: This variable has not been used.   [TINY_R_KIND]
src/gsi/read_gnssrspd.f90(41): remark #7712: This variable has not been used.   [HUGE_R_KIND]
src/gsi/read_gnssrspd.f90(42): remark #7712: This variable has not been used.   [EPS]
src/gsi/read_gnssrspd.f90(42): remark #7712: This variable has not been used.   [OMEPS]
src/gsi/read_gnssrspd.f90(42): remark #7712: This variable has not been used.   [RV]
src/gsi/read_gnssrspd.f90(42): remark #7712: This variable has not been used.   [GRAV]
src/gsi/read_gnssrspd.f90(47): remark #7712: This variable has not been used.   [ICSUBTYPE]
src/gsi/read_gnssrspd.f90(48): remark #7712: This variable has not been used.   [ITHIN_CONV]
src/gsi/read_gnssrspd.f90(48): remark #7712: This variable has not been used.   [RMESH_CONV]
src/gsi/read_gnssrspd.f90(48): remark #7712: This variable has not been used.   [PMESH_CONV]
src/gsi/read_gnssrspd.f90(49): remark #7712: This variable has not been used.   [PERTURB_OBS]
src/gsi/read_gnssrspd.f90(49): remark #7712: This variable has not been used.   [PERTURB_FACT]
src/gsi/read_gnssrspd.f90(51): remark #7712: This variable has not been used.   [AIRCRAFT_T_BC]
src/gsi/read_gnssrspd.f90(51): remark #7712: This variable has not been used.   [AIRCRAFT_T_BC_POF]
src/gsi/read_gnssrspd.f90(51): remark #7712: This variable has not been used.   [AIRCRAFT_T_BC_EXT]
src/gsi/read_gnssrspd.f90(52): remark #7712: This variable has not been used.   [ETABL]
src/gsi/read_gnssrspd.f90(53): remark #7712: This variable has not been used.   [ETABL_PS]
src/gsi/read_gnssrspd.f90(53): remark #7712: This variable has not been used.   [ISUBLE_PS]
src/gsi/read_gnssrspd.f90(53): remark #7712: This variable has not been used.   [MAXSUB_PS]
src/gsi/read_gnssrspd.f90(54): remark #7712: This variable has not been used.   [ETABL_Q]
src/gsi/read_gnssrspd.f90(54): remark #7712: This variable has not been used.   [ISUBLE_Q]
src/gsi/read_gnssrspd.f90(54): remark #7712: This variable has not been used.   [MAXSUB_Q]
src/gsi/read_gnssrspd.f90(55): remark #7712: This variable has not been used.   [ETABL_T]
src/gsi/read_gnssrspd.f90(55): remark #7712: This variable has not been used.   [ISUBLE_T]
src/gsi/read_gnssrspd.f90(55): remark #7712: This variable has not been used.   [MAXSUB_T]
src/gsi/read_gnssrspd.f90(56): remark #7712: This variable has not been used.   [ETABL_UV]
src/gsi/read_gnssrspd.f90(56): remark #7712: This variable has not been used.   [ISUBLE_UV]
src/gsi/read_gnssrspd.f90(56): remark #7712: This variable has not been used.   [MAXSUB_UV]
src/gsi/read_gnssrspd.f90(57): remark #7712: This variable has not been used.   [BTABL_PS]
src/gsi/read_gnssrspd.f90(58): remark #7712: This variable has not been used.   [BTABL_Q]
src/gsi/read_gnssrspd.f90(59): remark #7712: This variable has not been used.   [BTABL_T]
src/gsi/read_gnssrspd.f90(60): remark #7712: This variable has not been used.   [BTABL_UV]
src/gsi/read_gnssrspd.f90(61): remark #7712: This variable has not been used.   [IWINBGN]
src/gsi/read_gnssrspd.f90(61): remark #7712: This variable has not been used.   [THIN4D]
src/gsi/read_gnssrspd.f90(62): remark #7712: This variable has not been used.   [NJQC]
src/gsi/read_gnssrspd.f90(63): remark #7712: This variable has not been used.   [USE_ALL]
src/gsi/read_gnssrspd.f90(83): remark #7712: This variable has not been used.   [LTOB]
src/gsi/read_gnssrspd.f90(83): remark #7712: This variable has not been used.   [LQOB]
src/gsi/read_gnssrspd.f90(83): remark #7712: This variable has not been used.   [LUVOB]
src/gsi/read_gnssrspd.f90(83): remark #7712: This variable has not been used.   [LPSOB]
src/gsi/read_gnssrspd.f90(84): remark #7712: This variable has not been used.   [LUSE]
src/gsi/read_gnssrspd.f90(87): remark #7712: This variable has not been used.   [HDSTR]
src/gsi/read_gnssrspd.f90(87): remark #7712: This variable has not been used.   [SFMRSTR]
src/gsi/read_gnssrspd.f90(88): remark #7712: This variable has not been used.   [PSFSTR]
src/gsi/read_gnssrspd.f90(88): remark #7712: This variable has not been used.   [PRSSTR]
src/gsi/read_gnssrspd.f90(88): remark #7712: This variable has not been used.   [G10STR]
src/gsi/read_gnssrspd.f90(88): remark #7712: This variable has not been used.   [QCMSTR]
src/gsi/read_gnssrspd.f90(89): remark #7712: This variable has not been used.   [OBS_REGION]
src/gsi/read_gnssrspd.f90(93): remark #7712: This variable has not been used.   [BULSTR1]
src/gsi/read_gnssrspd.f90(93): remark #7712: This variable has not been used.   [BULSTR2]
src/gsi/read_gnssrspd.f90(94): remark #7712: This variable has not been used.   [OBSBUL]
src/gsi/read_gnssrspd.f90(100): remark #7712: This variable has not been used.   [KL]
src/gsi/read_gnssrspd.f90(100): remark #7712: This variable has not been used.   [K1]
src/gsi/read_gnssrspd.f90(100): remark #7712: This variable has not been used.   [K2]
src/gsi/read_gnssrspd.f90(100): remark #7712: This variable has not been used.   [J]
src/gsi/read_gnssrspd.f90(101): remark #7712: This variable has not been used.   [IRET]
src/gsi/read_gnssrspd.f90(101): remark #7712: This variable has not been used.   [LEVS]
src/gsi/read_gnssrspd.f90(108): remark #7712: This variable has not been used.   [ITHIN]
src/gsi/read_gnssrspd.f90(113): remark #7712: This variable has not been used.   [ITYPEY]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [IERR_PS]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [IERR_Q]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [IERR_T]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [IERR_UV]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [NCOUNT_PS]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [NCOUNT_Q]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [NCOUNT_T]
src/gsi/read_gnssrspd.f90(114): remark #7712: This variable has not been used.   [NCOUNT_UV]
src/gsi/read_gnssrspd.f90(116): remark #7712: This variable has not been used.   [P_QM]
src/gsi/read_gnssrspd.f90(116): remark #7712: This variable has not been used.   [G_QM]
src/gsi/read_gnssrspd.f90(116): remark #7712: This variable has not been used.   [T_QM]
src/gsi/read_gnssrspd.f90(116): remark #7712: This variable has not been used.   [Q_QM]
src/gsi/read_gnssrspd.f90(116): remark #7712: This variable has not been used.   [UV_QM]
src/gsi/read_gnssrspd.f90(116): remark #7712: This variable has not been used.   [PS_QM]
src/gsi/read_gnssrspd.f90(119): remark #7712: This variable has not been used.   [NLEVP]
src/gsi/read_gnssrspd.f90(120): remark #7712: This variable has not been used.   [PFLAG]
src/gsi/read_gnssrspd.f90(121): remark #7712: This variable has not been used.   [NTMP]
src/gsi/read_gnssrspd.f90(121): remark #7712: This variable has not been used.   [IIOUT]
src/gsi/read_gnssrspd.f90(122): remark #7712: This variable has not been used.   [KK]
src/gsi/read_gnssrspd.f90(122): remark #7712: This variable has not been used.   [KLON1]
src/gsi/read_gnssrspd.f90(122): remark #7712: This variable has not been used.   [KLAT1]
src/gsi/read_gnssrspd.f90(122): remark #7712: This variable has not been used.   [KLONP1]
src/gsi/read_gnssrspd.f90(122): remark #7712: This variable has not been used.   [KLATP1]
src/gsi/read_gnssrspd.f90(124): remark #7712: This variable has not been used.   [NMIND]
src/gsi/read_gnssrspd.f90(125): remark #7712: This variable has not been used.   [NIB]
src/gsi/read_gnssrspd.f90(127): remark #7712: This variable has not been used.   [IBIT]
src/gsi/read_gnssrspd.f90(147): remark #7712: This variable has not been used.   [RMESH]
src/gsi/read_gnssrspd.f90(149): remark #7712: This variable has not been used.   [TOE]
src/gsi/read_gnssrspd.f90(149): remark #7712: This variable has not been used.   [QOE]
src/gsi/read_gnssrspd.f90(149): remark #7712: This variable has not been used.   [PSOE]
src/gsi/read_gnssrspd.f90(149): remark #7712: This variable has not been used.   [VAR_JB]
src/gsi/read_gnssrspd.f90(153): remark #7712: This variable has not been used.   [U00]
src/gsi/read_gnssrspd.f90(153): remark #7712: This variable has not been used.   [V00]
src/gsi/read_gnssrspd.f90(153): remark #7712: This variable has not been used.   [U0]
src/gsi/read_gnssrspd.f90(153): remark #7712: This variable has not been used.   [V0]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [DX]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [DY]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [DX1]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [DY1]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [W00]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [W10]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [W01]
src/gsi/read_gnssrspd.f90(154): remark #7712: This variable has not been used.   [W11]
src/gsi/read_gnssrspd.f90(155): remark #7712: This variable has not been used.   [WDIR]
src/gsi/read_gnssrspd.f90(155): remark #7712: This variable has not been used.   [WSPD]
src/gsi/read_gnssrspd.f90(156): remark #7712: This variable has not been used.   [TOB]
src/gsi/read_gnssrspd.f90(156): remark #7712: This variable has not been used.   [UOB]
src/gsi/read_gnssrspd.f90(156): remark #7712: This variable has not been used.   [VOB]
src/gsi/read_gnssrspd.f90(156): remark #7712: This variable has not been used.   [QOB]
src/gsi/read_gnssrspd.f90(156): remark #7712: This variable has not been used.   [RROB]
src/gsi/read_gnssrspd.f90(157): remark #7712: This variable has not been used.   [RHOB]
src/gsi/read_gnssrspd.f90(157): remark #7712: This variable has not been used.   [TDOB]
src/gsi/read_gnssrspd.f90(158): remark #7712: This variable has not been used.   [POB_MB]
src/gsi/read_gnssrspd.f90(158): remark #7712: This variable has not been used.   [POB_CB]
src/gsi/read_gnssrspd.f90(158): remark #7712: This variable has not been used.   [POB_PA]
src/gsi/read_gnssrspd.f90(159): remark #7712: This variable has not been used.   [PSOB_MB]
src/gsi/read_gnssrspd.f90(159): remark #7712: This variable has not been used.   [PSOB_CB]
src/gsi/read_gnssrspd.f90(159): remark #7712: This variable has not been used.   [PSOB_PA]
src/gsi/read_gnssrspd.f90(160): remark #7712: This variable has not been used.   [QMAXERR]
src/gsi/read_gnssrspd.f90(161): remark #7712: This variable has not been used.   [DLNPOB]
src/gsi/read_gnssrspd.f90(161): remark #7712: This variable has not been used.   [PPB]
src/gsi/read_gnssrspd.f90(162): remark #7712: This variable has not been used.   [CRIT1]
src/gsi/read_gnssrspd.f90(162): remark #7712: This variable has not been used.   [TIMEDIF]
src/gsi/read_gnssrspd.f90(162): remark #7712: This variable has not been used.   [XMESH]
src/gsi/read_gnssrspd.f90(162): remark #7712: This variable has not been used.   [PMESH]
src/gsi/read_gnssrspd.f90(163): remark #7712: This variable has not been used.   [SSTIME]
src/gsi/read_gnssrspd.f90(165): remark #7712: This variable has not been used.   [ES]
src/gsi/read_gnssrspd.f90(165): remark #7712: This variable has not been used.   [QSAT]
src/gsi/read_gnssrspd.f90(165): remark #7712: This variable has not been used.   [RHOB_CALC]
src/gsi/read_gnssrspd.f90(165): remark #7712: This variable has not been used.   [TDOB_CALC]
src/gsi/read_gnssrspd.f90(165): remark #7712: This variable has not been used.   [TDRY]
src/gsi/read_gnssrspd.f90(166): remark #7712: This variable has not been used.   [DUMMY]
src/gsi/read_gnssrspd.f90(167): remark #7712: This variable has not been used.   [DEL]
src/gsi/read_gnssrspd.f90(167): remark #7712: This variable has not been used.   [EDIFF]
src/gsi/read_gnssrspd.f90(167): remark #7712: This variable has not been used.   [JBMIN]
src/gsi/read_gnssrspd.f90(168): remark #7712: This variable has not been used.   [TVFLG]
src/gsi/read_gnssrspd.f90(170): remark #7712: This variable has not been used.   [PRESL]
src/gsi/read_gnssrspd.f90(173): remark #7712: This variable has not been used.   [OBSTMP]
src/gsi/read_gnssrspd.f90(174): remark #7712: This variable has not been used.   [OBSWND]
src/gsi/read_gnssrspd.f90(175): remark #7712: This variable has not been used.   [OBSFMR]
src/gsi/read_gnssrspd.f90(176): remark #7712: This variable has not been used.   [OBSMST]
src/gsi/read_gnssrspd.f90(177): remark #7712: This variable has not been used.   [OBSPRS]
src/gsi/read_gnssrspd.f90(178): remark #7712: This variable has not been used.   [OBSPSF]
src/gsi/read_gnssrspd.f90(179): remark #7712: This variable has not been used.   [OBSG10]
src/gsi/read_gnssrspd.f90(180): remark #7712: This variable has not been used.   [OBSQCM]
src/gsi/read_gnssrspd.f90(187): remark #7712: This variable has not been used.   [PRESL_THIN]
src/gsi/setupgnssrspd.f90(53): remark #7712: This variable has not been used.   [AIRCRAFT_RECON]
src/gsi/setupgnssrspd.f90(120): remark #7712: This variable has not been used.   [DPRESAVE]
src/gsi/statsconv.f90(132): remark #7712: This variable has not been used.   [NREADGNSSRSPD]
src/gsi/statsconv.f90(132): remark #7712: This variable has not been used.   [NKEEPGNSSRSPD]
```
It's possible that not all the above variables are unused.  We need to check the above variables in the indicate routines to see if they are truly not used.  Those that are not used need to be removed.

KariA-Spire commented:
@RussTreadon-NOAA, I'll take a look at the list of used variables. From a quick glance there are several that should be used. 
@ilianagenkova, I will reply via email and Cc. @xincjin-NOAA

RussTreadon-NOAA committed:
remove unused variables from gnssrd codes (#628)

RussTreadon-NOAA committed:
remove more unused variables from gnssr code (#628)

RussTreadon-NOAA commented:
Remove most of the unused variables from gnssrwnd source code files.  The following variables are flagged as not being used
```
src/gsi/gsi_gnssrspdOper.F90(63): remark #7712: This variable has not been used.   [INIT_PASS]
src/gsi/gsi_gnssrspdOper.F90(63): remark #7712: This variable has not been used.   [LAST_PASS]
```
These variables are found in existing `gsi_*Oper.F90` files.  These variables will not, for the time being, be removed from `gsi_gnssrspdOper.F90`.

RussTreadon-NOAA commented:
**WCOSS2 ctests**
Install `KariA-Spire:gnssrwnd1.0` at 58aa1192 on Dogwood.  All tests pass.  This is an expected result.
```
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr747/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test #3: rrfs_3denvar_rdasens .............   Passed  731.83 sec
2/6 Test #6: global_enkf ......................   Passed  1030.57 sec
3/6 Test #2: rtma .............................   Passed  1088.81 sec
4/6 Test #5: hafs_3denvar_hybens ..............   Passed  1154.82 sec
5/6 Test #4: hafs_4denvar_glbens ..............   Passed  1213.79 sec
6/6 Test #1: global_4denvar ...................   Passed  1683.01 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) = 1683.02 sec
```

RussTreadon-NOAA commented:
@KariA-Spire , @emilyhcliu , and @xincjin-NOAA :  I would like to get this set of changes into `develop` sooner rather than later.  If Xin's tests with real data turn up problems, these problems can be addressed via new issues.  Is everyone OK with this approach?

xincjin-NOAA commented:
@RussTreadon-NOAA I am fine with the approach. Thanks,

ilianagenkova commented:
I'm fine with that approach too!

On Mon, Jul 8, 2024 at 12:36 PM xincjin-NOAA ***@***.***>
wrote:

> @RussTreadon-NOAA <https://github.com/RussTreadon-NOAA> I am fine with
> the approach. Thanks,
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/747#issuecomment-2214645000>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOC4YXW4HSG6ZECFTO4U3GTZLK5YRAVCNFSM6AAAAABHUSWSYWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMJUGY2DKMBQGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RussTreadon-NOAA commented:
Thank you @ilianagenkova and @xincjin-NOAA for your replies.  Although @KariA-Spire and @emilyhcliu have not yet replied, I will move forward and work with the GSI review committee to merge this PR into `develop`

RussTreadon-NOAA reviewed (approved):
Approve

KariA-Spire commented:
Thank you @RussTreadon-NOAA, @ilianagenkova, @xincjin-NOAA , @emilyhcliu!
